### ISSUE TITLE
Extract Smart Audio VTx code from BF Lua

### ISSUE THREAD

**Issue by teckel12** [teckel12/LuaTelemetry#157]
Goal is to strip out all the fluff from the BF Lua script that allows for Smart Audio VTx control that would provide a simple control interface for VTx control only. 

- Stripping out all PID stuff
- Stripping out all screen positioning for multiple controllers
- Stripping out the interface so it's just a simple function

Once this is done, a control panel or maybe directly from the config menu in Lua Telemetry control for VTx could be added. Or, at least a very stripped down independent Lua script could be created that provided just VTx control, saving a ton of limited memory.

I have no need for this myself as I always use the same channel, I also don't have any free UARTS. For these reasons, someone else needs to step up and do at least some of the work. I can't test anything, so it's almost impossible to program a working function without the ability to test.

So, I'm asking for help on this. It probably can't be done till someone offers assistance with the programming.

**Comment by teckel12** [comment]
https://github.com/teckel12/Taranis-VTx

This is a good start if someone wants to fork and take a look.  Strips out a lot, but it still uses tons of memory.

**Comment by teckel12** [comment]
The solution is to break the background and the display stuff out into multiple scripts and garbage collect after each.

**Comment by teckel12** [comment]
@davidngrc I know you can reduced the amount of memory used by selecting firmware options for OpenTX.  For example, turning off all build options except for the `noheli` switch.  I also know the OpenTX team is aware of the limited memory issue and they have increased memory available in the later releases.

But, there's a physical limitation to the memory in these transmitters.  OpenTX takes a lot, and they're always adding more features which use more memory.

**Comment by david-ng-hk** [comment]
I just add a feature request in OpenTx, hopping they would add the Virtual RAM.
ram in x9d sd card, since nowadays the SD Card is very fast (like 100MB/s for write).
Hope they will implement it.

**Comment by teckel12** [comment]
I'd bet the SD interface limits the speed to be much slower. Also, as there would be a lot of small-sized read/writes it would be significantly slower.  But still, the limited memory available does limit what you can do.

For a long time, free memory on the transmitter didn't serve a purpose, as there were not many scripts that had a lot of features to use it. But now with Lua Telemetry, BF Lua, and several others, using all available memory is a common problem. Heck, many people can't use the BF Lua script alone, let alone using it with other scripts.

**Comment by teckel12** [comment]
[Taranis VTx](https://github.com/teckel12/Taranis-VTx/releases/latest) released, enjoy!

**Comment by teckel12** [comment]
Reopened... Goal is to create a VTx control view to set VTx band, frequency, and transmitter power via the controller (PID control is out of scope for this project). 

VTx control could also be done from the config menu, as long as it doesn't use too much memory. Doing VTx control in its own view would make memory used mostly a non issue.

This will take some effort, as I'm not familiar with the Smart Audio protocol and MSP. Would like to make it compatible with both FrSky and TBS Crossfire. Will start with FrSky support and then look at adding Crossfire support.

This feature request has been re-opened after a generous donation of a VTx with Smart Audio (as well as an F722 SE to go with) from Matek. Now that I have the hardware, programming and testing will be possible.

If anyone has any Lua script programming info on MSP and Smart Audio, post links in this issue.

**Comment by teckel12** [comment]
Bad bot! This is going to be worked on after Crossfire support is released in v1.6.0

**Comment by laurentopia** [comment]
Thanks for this release!
self.fields[3].table = { 25, 200, 500, 800 }
Does this mean the powers are hard coded?

**Comment by teckel12** [comment]
@laurentopia Those are just the defaults for the display text.  Technically, it's really just power level 0, 1, 2, 3.  Different VTx have different actual power levels.  I have a Tiny Whoop like quad that has power levels 25, 100, 200.  So 0=25, 1=100, 2=200, 3=200.  So it's kinda confusing for sure.  But, there's no way to know what your actual VTx power levels are.

When I integrate this into Lua Telemetry, I may come up with a method of making this a bit more customizable.

**Comment by teckel12** [comment]
@laurentopia I'd post this as an issue in INAV as an enhancement. A developer may have interest and pick it up as a pet project.

**Comment by laurentopia** [comment]
Great, we'll see what happens.
